TSDB: Optimization: Merge postings using concrete type

bboreham opened the pull request:
Make `mergedPostings` generic, so we can call it with more specific types which is more efficient than making everything go through the `Postings` interface.

Move merge of head postings into index; remove the `Get` member which was not used for anything else, and fix up
tests.

~Unexport `PostingsFromDecbuf`; it looks like it should not have been exported.~

Preparatory work:
* Extend BenchmarkMemPostings_PostingsForLabelMatching to check merge speed.
* TSDB BenchmarkMerge: run fewer sizes.

Benchmark results:
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/tsdb/index
cpu: Intel(R) Core(TM) i7-14700K
                                                                   │ before.txt  │              after.txt              │
                                                                   │   sec/op    │    sec/op     vs base               │
Intersect/LongPostings1-28                                           10.70µ ± 0%    10.75µ ± 1%   +0.46% (p=0.026 n=6)
Intersect/LongPostings2-28                                           42.63m ± 2%    41.13m ± 1%   -3.53% (p=0.002 n=6)
Intersect/ManyPostings-28                                            111.9m ± 2%    111.0m ± 3%        ~ (p=0.699 n=6)
Merge/1-28                                                           193.5n ± 0%    177.8n ± 1%   -8.09% (p=0.002 n=6)
Merge/10-28                                                          9.947µ ± 0%   10.327µ ± 0%   +3.82% (p=0.002 n=6)
Merge/10000-28                                                       17.16m ± 2%    15.29m ± 3%  -10.92% (p=0.002 n=6)
Merge/100000-28                                                      241.8m ± 8%    199.3m ± 7%  -17.57% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=1000/matcher=fast-28     27.89µ ± 0%    27.08µ ± 2%   -2.91% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=1000/matcher=slow-28     4.894µ ± 4%    4.675µ ± 1%   -4.48% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=1000/matcher=all-28      176.7µ ± 2%    159.1µ ± 0%   -9.95% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=fast-28    282.2µ ± 1%    282.0µ ± 1%        ~ (p=0.818 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=slow-28    185.7m ± 2%    188.0m ± 3%        ~ (p=0.394 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=all-28     2.055m ± 7%    1.804m ± 2%  -12.25% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=fast-28   2.902m ± 1%    2.842m ± 1%   -2.07% (p=0.004 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=slow-28    2.371 ± 3%     2.331 ± 2%        ~ (p=0.240 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=all-28    24.17m ± 1%    21.35m ± 1%  -11.67% (p=0.002 n=6)
PostingStatsMaxHep-28                                                54.28m ± 0%    34.04m ± 0%  -37.28% (p=0.002 n=6)
geomean                                                              1.715m         1.586m        -7.49%

                                                                   │    before.txt    │               after.txt                │
                                                                   │       B/op       │     B/op       vs base                 │
Intersect/LongPostings1-28                                              224.0 ±  0%       224.0 ±  0%        ~ (p=1.000 n=6) ¹
Intersect/LongPostings2-28                                              224.0 ±  0%       224.0 ±  0%        ~ (p=1.000 n=6) ¹
Intersect/ManyPostings-28                                               32.00 ±  0%       32.00 ±  0%        ~ (p=1.000 n=6) ¹
Merge/1-28                                                              0.000 ±  0%       0.000 ±  0%        ~ (p=1.000 n=6) ¹
Merge/10-28                                                             784.0 ±  0%       560.0 ±  0%  -28.57% (p=0.002 n=6)
Merge/10000-28                                                        632.1Ki ±  0%     472.1Ki ±  0%  -25.31% (p=0.002 n=6)
Merge/100000-28                                                       6.109Mi ±  0%     4.578Mi ±  0%  -25.06% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=1000/matcher=fast-28      16.33Ki ±  0%     16.28Ki ±  0%   -0.29% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=1000/matcher=slow-28      16.02Ki ±  0%     16.02Ki ±  0%        ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=1000/matcher=all-28       128.1Ki ±  0%     104.1Ki ±  0%  -18.74% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=fast-28     160.4Ki ±  0%     160.3Ki ±  0%   -0.03% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=slow-28     356.2Ki ± 55%     355.9Ki ± 55%        ~ (p=0.143 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=all-28      1.242Mi ±  0%     1.008Mi ±  0%  -18.87% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=fast-28    1.532Mi ±  0%     1.532Mi ±  0%   -0.00% (p=0.002 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=slow-28    2.483Mi ± 23%     2.971Mi ±  0%        ~ (p=1.000 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=all-28    12.227Mi ±  0%     9.930Mi ±  0%  -18.79% (p=0.002 n=6)
PostingStatsMaxHep-28                                                   240.0 ±  0%       240.0 ±  0%        ~ (p=1.000 n=6) ¹
geomean                                                                             ²                   -7.72%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                                                   │   before.txt    │               after.txt                │
                                                                   │    allocs/op    │   allocs/op     vs base                │
Intersect/LongPostings1-28                                            6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
Intersect/LongPostings2-28                                            6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
Intersect/ManyPostings-28                                             1.000 ±   0%       1.000 ±   0%       ~ (p=1.000 n=6) ¹
Merge/1-28                                                            0.000 ±   0%       0.000 ±   0%       ~ (p=1.000 n=6) ¹
Merge/10-28                                                           3.000 ±   0%       3.000 ±   0%       ~ (p=1.000 n=6) ¹
Merge/10000-28                                                        3.000 ±   0%       3.000 ±   0%       ~ (p=1.000 n=6) ¹
Merge/100000-28                                                       3.000 ±   0%       3.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=1000/matcher=fast-28      6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=1000/matcher=slow-28      2.000 ±   0%       2.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=1000/matcher=all-28       6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=10000/matcher=fast-28     6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=10000/matcher=slow-28    2.042k ± 100%      2.032k ± 100%       ~ (p=0.147 n=6)
MemPostings_PostingsForLabelMatching/labels=10000/matcher=all-28      6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=100000/matcher=fast-28    6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
MemPostings_PostingsForLabelMatching/labels=100000/matcher=slow-28   6.129k ± 100%     12.186k ±   0%       ~ (p=1.000 n=6)
MemPostings_PostingsForLabelMatching/labels=100000/matcher=all-28     6.000 ±   0%       6.000 ±   0%       ~ (p=1.000 n=6) ¹
PostingStatsMaxHep-28                                                 1.000 ±   0%       1.000 ±   0%       ~ (p=1.000 n=6) ¹
geomean                                                                            ²                   +4.10%               ²
```

aknuds1 commented:
> Intersect/LongPostings1 slowdown seems consistent, although it was not changed in this PR. I will investigate further.

I also noticed this regression. Could you also rebase and fix the conflict(s)?

aknuds1 commented:
Got any further with investigating the `Intersect/LongPostings1` regression?

bboreham commented:
I have rebased and updated the benchmarks.  The regression seems to have gone away.

I can no longer apply the same optimisation to `Reader` methods, because the indirection added in #13567 means we don't know what type each postings is there.

aknuds1 commented on tsdb/index/postings_test.go:
```diff
@@ -392,8 +392,8 @@ func BenchmarkMerge(t *testing.B) {
-	its := make([]Postings, len(refs))
-	for _, nSeries := range []int{1, 10, 100, 1000, 10000, 100000} {
+	its := make([]*ListPostings, len(refs))
+	for _, nSeries := range []int{1, 10, 10000, 100000} {
```
Could you explain the reason for testing fewer `nSeries` here?

bboreham replied:
The intermediate ones don't tell us much, just slow down the process.

aknuds1 reviewed (approved):
LGTM, just see my question :)

bboreham commented:
I ran the TSDB `PostingsForMatchers` benchmarks, and see what looks like a large percentage slow-down in many cases.
However only the ones that take a couple of hundred nanoseconds slow down; all the ones that take longer speed up.
The reason is that the `Get` path did fewer allocations than the `Postings` path does now.

I could put back that case - check `len(values) == 1` in `Postings` - but I don't think it's worth it for 50ns per query.

```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/tsdb
cpu: Intel(R) Core(TM) i7-14700K
                                                                     │  before2.txt  │             after2.txt              │
                                                                     │    sec/op     │    sec/op     vs base               │
Querier/Head/PostingsForMatchers/n="1"-28                               203.8n ±  5%   229.6n ±  1%  +12.66% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="X"-28                               189.8n ±  1%   244.9n ±  2%  +29.06% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",j="foo"-28                       322.0n ±  1%   385.6n ±  1%  +19.74% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="X",j="foo"-28                       220.7n ±  7%   271.6n ±  0%  +23.09% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/j="foo",n="1"-28                       352.3n ±  1%   387.7n ±  1%  +10.03% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",j!="foo"-28                      342.3n ±  1%   378.8n ±  1%  +10.66% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i!="2"-28                        346.6n ±  1%   383.3n ±  1%  +10.60% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="X",j!="foo"-28                      225.1n ±  1%   262.7n ±  0%  +16.73% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i=~"1[0-9]",j=~"foo|bar"-28            679.6n ±  1%   463.0n ±  5%  -31.87% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/j=~"foo|bar"-28                        392.2n ±  2%   357.4n ±  0%   -8.86% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/j=~"XXX|YYY"-28                        314.3n ±  0%   283.0n ±  1%   -9.96% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/j=~"X.+"-28                            247.3n ±  1%   242.1n ±  0%   -2.08% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i=~"(1|2|3|4|5|6|20|55)"-28            590.2n ±  1%   395.1n ±  2%  -33.07% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i!~"(1|2|3|4|5|6|20|55)"-28            632.9n ±  1%   474.2n ±  2%  -25.06% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i=~"X|Y|Z"-28                          367.5n ±  1%   312.0n ±  0%  -15.11% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i!~"X|Y|Z"-28                          423.8n ±  0%   391.2n ±  1%   -7.68% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i=~".*"-28                             186.9n ±  1%   205.1n ±  0%   +9.74% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i=~"1.*"-28                            1.671m ±  6%   1.588m ±  5%   -4.98% (p=0.026 n=6)
Querier/Head/PostingsForMatchers/i=~".*1"-28                            1.044m ± 15%   1.084m ±  3%        ~ (p=0.310 n=6)
Querier/Head/PostingsForMatchers/i=~".+"-28                             3.966m ± 11%   3.302m ±  8%  -16.73% (p=0.004 n=6)
Querier/Head/PostingsForMatchers/i=~".+",j=~"X.+"-28                    3.865m ±  5%   3.410m ±  5%  -11.76% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i=~""-28                               3.986m ±  5%   3.383m ±  3%  -15.13% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/i!=""-28                               3.824m ±  8%   3.386m ± 10%  -11.45% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".*",j="foo"-28               375.6n ±  4%   413.1n ±  3%   +9.96% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="X",i=~".*",j="foo"-28               256.5n ±  1%   295.2n ±  0%  +15.09% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".*",i!="2",j="foo"-28        518.2n ±  2%   568.6n ±  1%   +9.73% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i!=""-28                         3.919m ± 10%   3.481m ±  7%  -11.16% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i!="",j="foo"-28                 3.905m ±  8%   3.377m ±  7%  -13.51% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"X.+"-28                3.852m ±  7%   3.537m ±  7%   -8.18% (p=0.009 n=6)
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"XXX|YYY"-28            3.927m ±  3%   3.812m ±  9%        ~ (p=0.093 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~"X|Y|Z",j="foo"-28            528.6n ±  2%   489.4n ±  2%   -7.42% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i!~"X|Y|Z",j="foo"-28            630.9n ±  1%   601.0n ±  1%   -4.74% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".+",j="foo"-28               3.965m ±  4%   3.457m ±  4%  -12.82% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~"1.+",j="foo"-28              1.629m ±  4%   1.623m ±  5%        ~ (p=0.699 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".*1.*",j="foo"-28            5.364m ±  5%   4.917m ±  4%   -8.33% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!="2",j="foo"-28        3.927m ±  8%   3.446m ±  5%  -12.24% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-28      5.784m ±  7%   5.493m ±  6%   -5.03% (p=0.026 n=6)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~".*2.*",j="foo"-28   10.065m ± 12%   8.482m ±  2%  -15.73% (p=0.002 n=6)
Querier/Head/PostingsForMatchers/n="X",i=~".+",i!~".*2.*",j="foo"-28    306.2n ±  1%   346.0n ±  2%  +13.00% (p=0.002 n=6)
geomean                                                                 15.43µ         14.86µ         -3.70%
```
